### PR TITLE
feat: implement new configtest command

### PR DESCRIPTION
**Pull request by snitin315**
**What kind of change does this PR introduce?**
feature

**Did you add tests for your changes?**
WIP
**If relevant, did you update the documentation?**
WIP
**Summary**

Syntax - `webpack configtest <path-to-config>`

**Does this PR introduce a breaking change?**
No

**Other information**
Fixes https://github.com/webpack/webpack-cli/issues/803

### CODE REVIEW

**Comment by snitin315** [comment]
Should we support validating multiple configs simultaneously like `webpack configtest <first-config> <second-config>`? 

Need to discuss /cc @webpack/cli-team

**Comment by snitin315** [comment]
I think it might be not possible for multiple configs as the process will exit as soon as an error config is found.

File: packages/configtest/src/index.ts [hunk 1]
```diff
@@ -0,0 +1,32 @@
+import { validate } from 'webpack';
+
+class ConfigTestCommand {
+    async apply(cli): Promise<void> {
+        const { logger } = cli;
+
+        await cli.makeCommand(
+            {
+                name: 'configtest',
```
Comment (alexander-akait) [inline packages/configtest/src/index.ts:9]: configtest <config-path>

Comment (alexander-akait) [reply packages/configtest/src/index.ts:9]: it is required arguments

File: packages/configtest/src/index.ts [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+                alias: 't',
+                description: 'Outputs information about your system.',
+                usage: '[configs]',
+                pkg: '@webpack-cli/configtest',
+            },
+            [],
+            async (program) => {
```
Comment (alexander-akait) [inline packages/configtest/src/index.ts:16]: (configPath, program)

**Review by alexander-akait — COMMENTED** [review]
Also please add this package to webpack-cli package.json (in peerDependenciesMeta) too

File: packages/configtest/src/index.ts [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+                    process.exit(2);
+                }
+
+                const { options } = await cli.resolveConfig({ config: [configPath] });
+
+                //eslint-disable-next-line @typescript-eslint/no-explicit-any
+                const validationErrors: any = validate(options);
+
+                if (validationErrors) {
+                    logger.error(validationErrors);
```
Comment (alexander-akait) [inline packages/configtest/src/index.ts:28]: process.exit(2) also please check how validate messages looks, we don't need stack

File: packages/configtest/src/index.ts [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+            async (configPath, program) => {
+                if (program.args.length > 1) {
+                    logger.error('Only one configuration can be validated at a time.');
+                    process.exit(2);
+                }
+
+                const { options } = await cli.resolveConfig({ config: [configPath] });
+
+                //eslint-disable-next-line @typescript-eslint/no-explicit-any
+                const validationErrors: any = validate(options);
```
Comment (alexander-akait) [inline packages/configtest/src/index.ts:25]: Ideally we should use webpack types here

File: packages/configtest/src/index.ts [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+                name: 'configtest <config-path>',
+                alias: 't',
+                description: 'Test your webpack configuration against validation errors.',
+                usage: '[configs]',
+                pkg: '@webpack-cli/configtest',
+            },
+            [],
+            async (configPath, program) => {
+                if (program.args.length > 1) {
+                    logger.error('Only one configuration can be validated at a time.');
```
Comment (alexander-akait) [inline packages/configtest/src/index.ts:18]: No need to check this, commander should output error self

File: packages/configtest/src/index.ts (hunk 3 above)
Comment (snitin315) [reply packages/configtest/src/index.ts:28]: WIP

Comment (snitin315) [reply packages/configtest/src/index.ts:23]: It's huge, how to reduce it?

Screenshot at 2021-01-05 17-45-35

Comment (alexander-akait) [reply packages/configtest/src/index.ts:23]: Logic here https://github.com/webpack/webpack-cli/blob/master/packages/webpack-cli/lib/webpack-cli.js#L1183

File: test/configtest/webpack.config.js [hunk 6]
```diff
@@ -0,0 +1,5 @@
+module.exports = {
+    mode: 'development',
+    target: 'node',
+    stats: 'verbose',
+};
```
Comment (alexander-akait) [inline test/configtest/webpack.config.js:5]: Add tests with broken JS syntax

**Review by alexander-akait — APPROVED** [review]
Good job

**Comment by alexander-akait** [comment]
Feel free to covert it to completed PR from Draft, I think we can merge it

**Comment by snitin315** [comment]
Let's see how CI run goes.

**Comment by snitin315** [comment]
Ah, need some tweaks for webpack-4

File: packages/configtest/src/index.ts [hunk 7]
```diff
@@ -13,22 +15,29 @@ class ConfigTestCommand {
...
-
-                if (error) {
-                    logger.error(isValidationError(error) ? error.message : error);
+                const error: any = validate(options);
+
+                if (error && error.length) {
+                    if (isValidationError(error)) {
+                        // eslint-disable-next-line @typescript-eslint/ban-ts-ignore
+                        // @ts-ignore
+                        logger.error(isWebpack5 ? new ValidationError(error) : new WebpackOptionsValidationError(error));
```
Comment (alexander-akait) [inline packages/configtest/src/index.ts:37]: Why?

Comment (snitin315) [reply packages/configtest/src/index.ts:37]: else it logs - 

Screenshot at 2021-01-05 19-26-58

`migrate` uses the same logic https://github.com/webpack/webpack-cli/blob/19a25cf83dc2f680a5028f4b449d7f79895231f0/packages/migrate/src/index.ts#L133

Comment (alexander-akait) [reply packages/configtest/src/index.ts:37]: weird, I will look at this in near future

File: packages/webpack-cli/README.md [hunk 8]
```diff
@@ -63,15 +63,16 @@ npx webpack-cli --help verbose
...
+  bundle | b      Run webpack
+  help | h        Display help for commands and options
+  version | v     Output version number of the 'webpack', 'webpack-cli' and other related packages
+  init | c        Initialize a new webpack configuration
+  migrate | m     Migrate a configuration to a new version
+  loader | l      Scaffold a loader repository
+  plugin | p      Scaffold a plugin repository
+  info | i        Outputs information about your system and dependencies
+  serve | s       Run the webpack Dev Server
+  configtest | t  Run the webpack Dev Server
```
Comment (alexander-akait) [inline packages/webpack-cli/README.md:75]: Something wrong with description

Comment (snitin315) [reply packages/webpack-cli/README.md]: fixed

File: packages/configtest/src/index.ts [hunk 9]
```diff
@@ -0,0 +1,50 @@
+import { validate, version, ValidationError, WebpackOptionsValidationError } from 'webpack';
+
+const isWebpack5: boolean = version.startsWith('5');
```
[the remark this answers is not in the document]
Comment (snitin315) [reply packages/configtest/src/index.ts]: fixed

File: packages/configtest/src/index.ts [hunk 10]
```diff
@@ -0,0 +1,50 @@
...
+
+                if (error && error.length) {
+                    if (isWebpack5) {
+                        logger.error(isValidationError(error) ? error.message : error);
+                    } else {
+                        // eslint-disable-next-line @typescript-eslint/ban-ts-ignore
+                        // @ts-ignore
+                        logger.error(new WebpackOptionsValidationError(error));
+                    }
+                    process.exit(2);
```
Comment (alexander-akait) [inline packages/configtest/src/index.ts:41]: It should be:
```
if (isValidationError) {
  logger.error(error.message)
} else {
  logger.error(error)
}

process.exit(2);
```

**Review by alexander-akait — CHANGES_REQUESTED** [review]
Need small improve

File: OPTIONS.md [hunk 11]
```diff
@@ -559,14 +559,15 @@ Global options:
...
-  help|h                                     Display help for commands and options.
   version|v                                  Output the version number of 'webpack', 'webpack-cli' and 'webpack-dev-server' and commands.
+  help|h                                     Display help for commands and options.
   serve|s [options]                          Run the webpack dev server.
   info|i [options]                           Outputs information about your system.
   init|c [options] [scaffold...]             Initialize a new webpack configuration.
   loader|l [output-path]                     Scaffold a loader.
-  plugin|p [output-path]                     Scaffold a plugin.
   migrate|m <config-path> [new-config-path]  Migrate a configuration to a new version.
+  configtest|t <config-path>                 Test your webpack configuration against validation errors.
```
Comment (rishabh3112) [inline OPTIONS.md:569]:
```suggestion
  configtest|t <config-path>                 Tests webpack configuration against validation errors.
```

File: packages/configtest/src/index.ts [hunk 12]
```diff
@@ -0,0 +1,50 @@
...
+                }
+
+                logger.success('There are no errors in the given webpack configuration.');
```
Comment (rishabh3112) [inline packages/configtest/src/index.ts:44]:
```suggestion
                logger.success('There are no validation errors in the given webpack configuration.');
```

File: packages/webpack-cli/README.md [hunk 13]
```diff
@@ -63,15 +63,16 @@ npx webpack-cli --help verbose
...
+  bundle | b      Run webpack
+  help | h        Display help for commands and options
+  version | v     Output version number of the 'webpack', 'webpack-cli' and other related packages
+  init | c        Initialize a new webpack configuration
+  migrate | m     Migrate a configuration to a new version
+  loader | l      Scaffold a loader repository
+  plugin | p      Scaffold a plugin repository
+  info | i        Outputs information about your system and dependencies
+  serve | s       Run the webpack Dev Server
+  configtest | t  Test your webpack configuration against validation errors.
```
Comment (rishabh3112) [inline packages/webpack-cli/README.md:75]:
```suggestion
  configtest | t  Tests webpack configuration against validation errors.
```

File: packages/configtest/package.json [hunk 14]
```diff
@@ -0,0 +1,18 @@
+{
+  "name": "@webpack-cli/configtest",
+  "version": "1.0.0",
+  "description": "Test your webpack configuration against validation errors.",
```
Comment (rishabh3112) [inline packages/configtest/package.json:4]:
```suggestion
  "description": "Tests webpack configuration against validation errors.",
```

File: packages/configtest/src/index.ts [hunk 15]
```diff
@@ -0,0 +1,50 @@
+class ConfigTestCommand {
+    async apply(cli): Promise<void> {
+        const { logger } = cli;
+
+        await cli.makeCommand(
+            {
+                name: 'configtest <config-path>',
+                alias: 't',
+                description: 'Test your webpack configuration against validation errors.',
+                usage: '[configs]',
```
Comment (rishabh3112) [inline packages/configtest/src/index.ts:10]:
```suggestion
                usage: '[config]',
```

**Review by rishabh3112 — COMMENTED** [review]
Looks good, left few suggestions.

File: packages/configtest/src/index.ts (hunk 15 above)
Comment (jamesgeorge007) [inline packages/configtest/src/index.ts:10]:
```suggestion
                usage: '<config-path>',
```

File: packages/configtest/src/index.ts (hunk 10 above)
Comment (snitin315) [reply packages/configtest/src/index.ts:41]: It's not working with `webpack-4`

File: packages/configtest/package.json (hunk 14 above)
Comment (snitin315) [reply packages/configtest/package.json:4]: I'm not sure `Tests` will be suitable here. I replaced `your` with `any`. 
`Test any webpack configuration against validation errors.`

File: packages/configtest/src/index.ts [hunk 16]
```diff
@@ -0,0 +1,50 @@
...
+                    //eslint-disable-next-line @typescript-eslint/no-explicit-any
+                    const webpackValidationError: any = ValidationError || WebpackOptionsValidationError;
+
+                    return error instanceof webpackValidationError;
+                };
+
+                //eslint-disable-next-line @typescript-eslint/no-explicit-any
+                const error: any = validate(options);
+
+                if (error && error.length) {
```
[the remark this answers is not in the document]
Comment (snitin315) [reply packages/configtest/src/index.ts:33]: fixed

File: packages/configtest/src/index.ts (hunk 12 above)
Comment (snitin315) [reply packages/configtest/src/index.ts]: resolved

File: packages/configtest/package.json (hunk 14 above)
Comment (alexander-akait) [reply packages/configtest/package.json]: It should be `Tests`

Comment (snitin315) [reply packages/configtest/package.json]: fixed

**Pull request merged**